print-macsec.c

```diff
@@ -0,0 +1,208 @@
+ *     composed of a MAC address + 16-bit port number
+ */
+struct macsec_sectag {
+	u_char  tci_an;
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+	u_char short_length:6,
+		     unused:2;
+#else /* __BYTE_ORDER == __BIG_ENDIAN */
+	u_char        unused:2,
+		short_length:6;
+#endif
+	uint32_t packet_number;
+	u_char secure_channel_id[8]; /* optional */
+} __attribute__((packed));
+
+/* IEEE 802.1AE-2006 9.5 */
+#define MACSEC_TCI_VERSION 0x80
+#define MACSEC_TCI_ES      0x40 /* end station */
+#define MACSEC_TCI_SC      0x20 /* SCI present */
+#define MACSEC_TCI_SCB     0x10 /* epon */
+#define MACSEC_TCI_E       0x08 /* encryption */
+#define MACSEC_TCI_C       0x04 /* changed text */
+#define MACSEC_AN_MASK     0x03 /* association number */
+#define MACSEC_TCI_FLAGS   (MACSEC_TCI_ES | MACSEC_TCI_SC | MACSEC_TCI_SCB | MACSEC_TCI_E | MACSEC_TCI_C)
+#define MACSEC_TCI_CONFID  (MACSEC_TCI_E | MACSEC_TCI_C)
+
+#define MACSEC_SECTAG_LEN_NOSCI 6
+#define MACSEC_SECTAG_LEN_SCI 14
+static int
+ieee8021ae_sectag_len(const struct macsec_sectag *sectag)
+{
+	return (sectag->tci_an & MACSEC_TCI_SC) ?
+	       MACSEC_SECTAG_LEN_SCI :
+	       MACSEC_SECTAG_LEN_NOSCI;
+}
+
+static int macsec_check_length(const struct macsec_sectag *sectag, u_int length, u_int caplen)
+{
+	u_int len;
+
+	/* we need the full MACsec header in the capture */
+	if (caplen < (MACSEC_SECTAG_LEN_NOSCI + 2))
+		return 0;
+
+	len = ieee8021ae_sectag_len(sectag);
+	if (caplen < (len + 2))
+		return 0;
+
+	if (sectag->short_length != 0) {
+		/* original packet must have exact length */
+		u_int exact = ETHER_HDRLEN + len + 2 + sectag->short_length;
+		return exact == length;
+	} else {
+		/* original packet must not be short */
+		u_int minlen = ETHER_HDRLEN + len + 2 + 48;
+		return length >= minlen;
+	}
+
+	return 1;
+}
+
+#define SCI_FMT "%016" PRIx64
+
+static const struct tok macsec_flag_values[] = {
+	{ MACSEC_TCI_E,   "E" },
+	{ MACSEC_TCI_C,   "C" },
+	{ MACSEC_TCI_ES,  "S" },
+	{ MACSEC_TCI_SCB, "B" },
+	{ MACSEC_TCI_SC,  "I" },
+	{ 0, NULL }
+};
+
+/* returns < 0 iff the packet can be decoded completely */
+int macsec_print(netdissect_options *ndo, const u_char **bp,
+		 u_int *lengthp, u_int *caplenp, u_int *hdrlenp,
+		 u_short *length_type)
+{
+	const u_char *p = *bp;
+	u_int length = *lengthp;
+	u_int caplen = *caplenp;
+	u_int hdrlen = *hdrlenp;
+	const struct macsec_sectag *sectag = (const struct macsec_sectag *)p;
+	u_int len;
+
+	if (!macsec_check_length(sectag, length, caplen)) {
+		ND_PRINT((ndo, tstr));
+		return hdrlen + caplen;
+	}
+
+	if (sectag->unused || sectag->tci_an & MACSEC_TCI_VERSION) {
+		ND_PRINT((ndo, "%s", istr));
+		return hdrlen + caplen;
+	}
+
+	if (ndo->ndo_eflag) {
+		char buf[128];
+		int n = snprintf(buf, sizeof(buf), "an %d, pn %d, flags %s",
+				 sectag->tci_an & MACSEC_AN_MASK,
+				 EXTRACT_32BITS(&sectag->packet_number),
+				 bittok2str_nosep(macsec_flag_values, "none",
+						  sectag->tci_an & MACSEC_TCI_FLAGS));
```

Comment:
And use `GET_U_1()` here as well - `GET_U_1(sectag->tci_an) & MACSEC_TCI_FLAGS`.